### PR TITLE
Enable autofix on multiline maths

### PR DESCRIPTION
**Pull request by jlelong**
Related to #499.

Autofix `$ $` to `\( \)` and `$$ $$` to `\[ \]` works for multiline expressions

### CODE REVIEW

File: src/providers/codeactions.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import * as vs from 'vscode'
 
 import type { Extension } from '../main'
+import { TeXMathEnvFinder } from '../providers/preview/mathpreviewlib/texmathenvfinder'
```
Comment (tamuratak) [inline src/providers/codeactions.ts:4]: We should not directly import `mathpreviewlib/texmathenvfinder`.

Please make `MathPreview.texMathEnvFinder` public and use `Extension.mathPreview.texMathEnvFinder`.

https://github.com/James-Yu/LaTeX-Workshop/blob/e3c795752b10d0f6dd56df9280002b16bad8b57f/src/providers/preview/mathpreview.ts#L19-L29

**Review by tamuratak — CHANGES_REQUESTED** [review]

**Comment by jlelong** [comment]
I have made the necessary changes. For the sake of coherence, I have move all the `replaceXXX` functions inside the `CodeActions` class.

File: src/providers/codeactions.ts (hunk 1 above)
Comment (jlelong) [reply src/providers/codeactions.ts]: I have made the necessary changes. For the sake of coherence, I have move all the `replaceXXX` functions inside the `CodeActions` class.

File: src/providers/codeactions.ts [hunk 2]
```diff
@@ -172,16 +132,64 @@ export class CodeActions {
...
-                replaceRangeWithString(document, range, fixString)
+                this.replaceRangeWithString(document, range, fixString)
                 break
             case 45:
-                replaceMathDelimitersInRange(document, range, '$$', '\\[', '\\]')
+                this.replaceMathDelimitersInRange(document, range, '$$', '\\[', '\\]')
                 break
             case 46:
-                replaceMathDelimitersInRange(document, range, '$', '\\(', '\\)')
+                this.replaceMathDelimitersInRange(document, range, '$', '\\(', '\\)')
```
Comment (tamuratak) [inline src/providers/codeactions.ts:141]: We should not call `texMathEnvFinder` for `$`. It does not work.

**Review by tamuratak — CHANGES_REQUESTED** [review]

**Comment by tamuratak** [comment]
This PR does not work on some rare cases. For example:

```latex
$ x \$ $
```

```latex
$$
x
$$$y$
```

It would be appropriate to use `Extension.pegParser.parseLatex` instead of `texMathEnvFinder`. I don't strongly push using `parseLatex`.

File: src/providers/codeactions.ts (hunk 2 above)
Comment (jlelong) [reply src/providers/codeactions.ts:141]: For standard expressions like

```latex
Some maths $1 + 3 =
4$ within text
```

 it is working. Can you give a failing example?

Comment (tamuratak) [reply src/providers/codeactions.ts:141]: I mean, it is not intended behavior. The behavior might change in the future refactoring.

**Comment by jlelong** [comment]
Actually, `texMathEnvFinder.findEndPair` should use `Extension.pegParser.parseLatex` and only use the current implementation as a fallback. Let's keep it for an other PR

File: src/providers/codeactions.ts (hunk 2 above)
Comment (jlelong) [reply src/providers/codeactions.ts:141]: Fine, we can drop the support for `$...$` split over several lines. This is consistent with the hovering feature.

**Review by tamuratak — APPROVED** [review]

File: src/providers/codeactions.ts [hunk 3]
```diff
@@ -172,16 +132,68 @@ export class CodeActions {
...
+        const edit = new vs.WorkspaceEdit()
+        edit.replace(document.uri, range, replacementString)
+        return vs.workspace.applyEdit(edit)
+    }
+
+    private replaceRangeWithRepeatedString(document: vs.TextDocument, range: vs.Range, replacementString: string) {
+        return this.replaceRangeWithString(document, range, replacementString.repeat(range.end.character - range.start.character))
+    }
+
+    private replaceMathDelimitersInRange(document: vs.TextDocument, range: vs.Range, oldDelim: string, startDelim: string, endDelim: string) {
```
Comment (tamuratak) [inline src/providers/codeactions.ts:172]: It would be better to change the type of `oldDelim` from `string` to `'$' | '$$'`.

**Review by tamuratak — CHANGES_REQUESTED** [review]

**Comment by jlelong** [comment]
```latex
$$
x
$$$y$
```
works fine now.

**Review by tamuratak — APPROVED** [review]

**Pull request merged**